### ISSUE TITLE
allowBlockStart option in lines-around-comment doesn't support if blocks

### ISSUE THREAD

**Issue by paglias** [eslint/eslint#4339]
Code like this:

```
if (cond) {
  // a comment at the beginning
}
```

with:

`/*eslint lines-around-comment: [2, { "beforeLineComment": true, "allowBlockStart": true }]*/`

shouldn't throw an error as it doesn't on function blocks

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by ilyavolodin** [comment]
This is a good question. This rule was written for ES5, and in ES5 `if` is not a block. I'm not sure we should update it for ES6. @eslint/eslint-team

**Comment by alberto** [comment]
@ilyavolodin if create block statements and are already supported by this rule, see [test case](https://github.com/eslint/eslint/blob/master/tests/lib/rules/lines-around-comment.js#L125-L129).

It seems to fail when there is only a comment in the body.

**Comment by ilyavolodin** [comment]
Ahh... Ok, in that case we should fix that.

**Comment by alberto** [comment]
The issue is the last ancestor is the `IfStatement` instead of the `BlockStatement` when the block is empty. Is that a bug in estraverse?

**Comment by paglias** [comment]
Thanks for looking at this! I didn't notice this was limited to comment-only if blocks, I just got the error and thought it was general

**Comment by nzakas** [comment]
@alberto no. Comments are not considered part of the AST. We hack it so our traversal includes comments, but the comment is not technically a child of the block.

**Comment by alberto** [comment]
And where is the hack done? From what I saw, I guess it's done through the `comment-event-generator` but I couldn't find the subscriber.

**Comment by nzakas** [comment]
ESLint works on an `EventEmitter`. Whatever is exported from a rule (like `Identifier`, `BlockComment`, etc.) is automatically subscribed to that `EventEmitter` when the rule is loaded.

**Comment by alberto** [comment]
Maybe I'm missing something, but from the debugging I made, estraverse somehow is also aware of the comments. `context.getAncestors();` (which delegates to `estraverse.parents()`) returns the ancestors of a given comment node and returns a different last ancestor depending on whether the if body is empty or not (`IfStatement` vs `BlockStatement`). I don't think that should be the case, should it?

Otherwise, it could be fixed simply by adding a check for `isCommentAtParentStart(node, "IfStatement")`.

**Comment by nzakas** [comment]
estraverse is not aware of comments. The functionality of `estraverse.parents()` just uses whatever the last node it traversed to as the base. It doesn't traverse to a comment because comments are not formally stored in the AST.
